### PR TITLE
Name more chunks

### PR DESCRIPTION
**Pull request by oparisy**
I made a pass on the chunks trees generated by `parse-3ds` on some freely available models, and added names for all "Unknown" ones I encountered. My references were mostly  http://www.fileformat.info/format/3ds/egff.htm, http://personales.unican.es/camareroc/3ds/3dsformat.pdf and http://www.martinreddy.net/gfx/3d/3DS.spec. I tried to add as clear names a possible.

I also took the opportunity to parse materials name.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -72,6 +73,19 @@ var CHUNK_NAMES = {
...
 
+function parseMaterialNameChunk(buf) {
+  return {
+    materialName: fromAsciiz(buf)
+  };
+}
+
+
+function fromAsciiz(buf) {
+  // TODO What would be the proper encoding?
```
Comment (thibauts) [inline index.js]: utf-8 probably

File: index.js [hunk 2]
```diff
@@ -72,6 +73,19 @@ var CHUNK_NAMES = {
...
 
 
+function parseMaterialNameChunk(buf) {
+  return {
+    materialName: fromAsciiz(buf)
+  };
+}
+
+
+function fromAsciiz(buf) {
```
Comment (thibauts) [inline index.js]: Could this function be made to work in `parseObjectChunk` ? If the source string really is an ASCIIZ string then its length will safely be usable as the buffer offset in the aforementioned function.

**Comment by thibauts** [comment]
I commented directly on the file [here](https://github.com/thibauts/parse-3ds/pull/3/commits/8f422d5e7fb572ae7f215d675fc3a9b203616769).

As an aside, if you use this module you may want to recover the lost bits in face parsing to get smoothing groups back.

**Comment by oparisy** [comment]
I am not clear on what you are suggesting wrt to the ASCIIZ part.

`parseMaterialNameChunk` and `fromAsciiz` are simple because in the case of the material name, I know the chunk `data` buffer is uniquely and entirely composed of characters.

I did not notice the ASCIIZ part of `parseObjectChunk`, and I see that it is safer. Are you suggesting I could factor out the beginning of `parseObjectChunk` in `fromAsciiz`?

**Comment by thibauts** [comment]
Yes. That may be nice to have a single function to extract them from buffers as this seems to be something common in the file format.

**Comment by oparisy** [comment]
With respect to encoding: have you reasons to believe encoding is UTF-8 based? I have no access to 3DS to perform a test, and the Blender exporter simply converts non-ascii characters to `?`.

The various readers I've had a look at either use `ASCII` of `ISO-8859-1`, which are both single-byte encoding. With UTF-8, we cannot compute the offset as `string length + 1` anymore. This complicates `fromAsciiz` signature, which will have to return the string and the offset separately.

**Comment by thibauts** [comment]
No I'm pretty sure the encoding is basic US-ASCII in most cases (for serializers following the specs), but `buf.toString()` takes as a parameter the _output_ encoding as I understand it. If I'm not mistaken UTF-8 is backwards compatible with US-ASCII and each input byte will be translated to exactly one character.

This would break if the input is iso-8859-1 and mess up some characters (utf-8 not backward compatible), but `buf.toString('ascii')` would mess them up too, as per the node docs the high bit is cleared on conversion. You could use the `binary` encoding which will work with both US-ASCII and ISO-8859-1 though.

That being said I'm not totally clear with the node documentation as it seems to mix internal storage format with with various ASCII-printable representations of raw data.

I guess to know better if this param sometimes represents how the code will interpret the source data or not, we should directly check the node source.

**Comment by oparisy** [comment]
By the JavaScript standard, node.js must expose its characters as some variant of UCS-2, and probably uses UTF-16 internally (source: [mathiasbynens.be](https://mathiasbynens.be/notes/javascript-encoding)). We have no access to this internal representation, but we can be confident in its capability to store any ISO-8859-1 character we'll throw at it.

The encoding provided to `buffer.toString` is used to interpret the bytes stored in the `buffer`. I do not believe their final, internal storage after interpretation is relevant (since we have no access to them), as long as they can express the interpreted characters as stated above.

I just wrote a snippet using ISO-88859-1 to demonstrate this behavior, but it seems like this encoding [is not supported by buffers](http://stackoverflow.com/a/14551669/38096), and would require an extra dependency.

Well, since we have no sample of 8 bits or more names in 3DS files, any ASCII (7 bits) compatible encoding will do anyway. But as you describe we would need something that degrade nicely in the presence of 8 bits characters (I just tested the hi-bit stripping behavior of ASCII and find it confusing).

Or find someone owning 3DS and ask him for a file with an accentuated material name :)

**Comment by oparisy** [comment]
Mostly this boils down to the fact that we have no precise specification of strings encoding nor examples of 3ds files embedding non-ASCII characters, so I'd suggest the following:
- accept an optional encoding (since PR #2 provided us with a mechanism to do so),
- default to the ASCII (7 bits) encoding when none is specified, since we know how it will behave in the presence of 8 bits chars,
- document that, should anyone need to specify the encoding, it would be nice of him to create an issue so that this could be investigated in more details.

I do not think we can do more with current knowledge.

Would you be OK with this?

**Comment by thibauts** [comment]
Regarding ISO-8859-1, the node documentation says you can have it with the "binary" encoding (https://nodejs.org/api/buffer.html#buffer_buffers_and_character_encodings). My guess is it doesn't touch the bytes and copies them as-is.

After reading your comments I looked into the node and v8 source because this encoding thing confuses me. Things proceed as follows if we follow the path of an UTF-8 string :
- toString (https://github.com/nodejs/node/blob/master/lib/buffer.js#L506-L516)
- slowToString (https://github.com/nodejs/node/blob/master/lib/buffer.js#L440-L503)
- Utf8Slice (https://github.com/nodejs/node/blob/master/src/node_buffer.cc#L532-L534)
- StringSlice (https://github.com/nodejs/node/blob/master/src/node_buffer.cc#L459-L473)
- StringBytes::Encode (https://github.com/nodejs/node/blob/master/src/string_bytes.cc#L770-L868)

Here it appears that node _encodes_ for every encoding, except for the utf-8 case where it actually _decodes_ the input. This makes sense practically speaking but still confuses me as to what `Encode` really means here.

But well.

Anyway the `binary` encoding simply copies the bytes (https://github.com/nodejs/node/blob/master/src/string_bytes.cc#L818-L823) so it would read both US-ASCII and ISO-8859-1. Ultimately both will be stored as a v8 "external one byte string" and supposedly handled as such by node.

So I'd use `binary` as a default as it should work in most cases, and allow an `encoding` parameter in options that accepts ascii (in case we need to force it?), utf16le / ucs2 and utf8.

But I may be missing something. Encodings are hell ^^

**Comment by oparisy** [comment]
I didn't notice `binary`. Nice catch!

I updated this PR following our discussion. An explicit `encoding` option can now be specified, and defaults to `binary`. Things work fine with the files I tested this on.

**Comment by thibauts** [comment]
Published to npm as 1.1.3. Thanks :)

And sorry for the delay !

**Comment by oparisy** [comment]
Great! Thanks.

**Pull request merged**